// @microsoft/live-share/bin/internals/consts.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the Microsoft Live Share SDK License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSynchronizerEvents = exports.TelemetryEvents = void 0;
/**
 * List of telemetry events.
 * @hidden
 * @remarks
 * Wrap with a call to transmit() if the event should be transmitted to the telemetry service.
 */
exports.TelemetryEvents = {
    LivePresence: {
        LocalPresenceChanged: "LivePresence:LocalPresenceChange",
        RemotePresenceChanged: "LivePresence:RemotePresenceChange",
        GetClientInfoError: "LivePresence:GetClientInfoError",
        RoleVerificationError: "LivePresence:RoleVerificationError",
    },
    LiveState: {
        StateChanged: "LiveState:StateChanged",
        RoleVerificationError: "LiveState:RoleVerificationError",
    },
};
/**
 * @hidden
 */
function transmit(eventName) {
    return `${eventName}#transmit`;
}
/**
 * @hidden
 */
exports.ObjectSynchronizerEvents = {
    update: "update",
    connect: "connect",
};
